src/query/v2/plan/operator_distributed.cpp

```diff
@@ -595,4 +597,149 @@ void Produce::ProduceCursor::Shutdown() { input_cursor_->Shutdown(); }
+            frame_idx_ = *frame_idx_ + 1;
+          }
+          return true;
+        }
+        frame_idx_ = *frame_idx_ + 1;
+      }
+
+      frame_idx_ = std::nullopt;
+    }
+  }
+
+  void Shutdown() override { input_cursor_->Shutdown(); }
+
+  void Reset() override {
+    input_cursor_->Reset();
+    input_value_.clear();
+    input_value_it_ = input_value_.end();
+  }
+
+ private:
+  const Unwind &self_;
+  const UniqueCursorPtr input_cursor_;
+  // typed values we are unwinding and yielding
+  utils::pmr::vector<TypedValue> input_value_;
+  // current position in input_value_
+  decltype(input_value_)::iterator input_value_it_ = input_value_.end();
+  std::optional<size_t> frame_idx_{std::nullopt};
+  size_t populated_frames_{0};
+};
+
+UniqueCursorPtr Unwind::MakeCursor(utils::MemoryResource *mem) const {
+  EventCounter::IncrementCounter(EventCounter::UnwindOperator);
+
+  return MakeUniqueCursorPtr<UnwindCursor>(mem, *this, mem);
+}
+
+// Returns boolean result of evaluating filter expression. Null is treated as
+// false. Other non boolean values raise a QueryRuntimeException.
+bool EvaluateFilter(ExpressionEvaluator &evaluator, Expression *filter) {
+  TypedValue result = filter->Accept(evaluator);
+  // Null is treated like false.
+  if (result.IsNull()) return false;
+  if (result.type() != TypedValue::Type::Bool)
+    throw QueryRuntimeException("Filter expression must evaluate to bool or null, got {}.", result.type());
+  return result.ValueBool();
+}
+
+Filter::Filter(const std::shared_ptr<LogicalOperator> &input, Expression *expression)
+    : input_(input ? input : std::make_shared<Once>()), expression_(expression) {}
+
+ACCEPT_WITH_INPUT(Filter)
+
+UniqueCursorPtr Filter::MakeCursor(utils::MemoryResource *mem) const {
+  EventCounter::IncrementCounter(EventCounter::FilterOperator);
+
+  return MakeUniqueCursorPtr<FilterCursor>(mem, *this, mem);
+}
+
+std::vector<Symbol> Filter::ModifiedSymbols(const SymbolTable &table) const { return input_->ModifiedSymbols(table); }
+
+Filter::FilterCursor::FilterCursor(const Filter &self, utils::MemoryResource *mem)
+    : self_(self), input_cursor_(self_.input_->MakeCursor(mem)) {}
+
+bool Filter::FilterCursor::Pull(MultiFrame &multiframe, ExecutionContext &context) {
+  SCOPED_PROFILE_OP("Filter");
+
+  while (input_cursor_->Pull(multiframe, context)) {
+    // TODO(gitbuda): Make frames invalid only there is something to return because in the case with many pull without
+    // return multiframe will be in a "polluted" state.
+    for (auto idx = 0; idx < multiframe.Size(); ++idx) {
+      auto &frame = multiframe.GetFrame(idx);
+      if (!frame.IsValid()) {
+        continue;
+      }
+      ExpressionEvaluator evaluator(&frame, context.symbol_table, context.evaluation_context, context.db_accessor,
+                                    storage::v3::View::OLD);
+      if (!EvaluateFilter(evaluator, self_.expression_)) {
+        frame.MakeInvalid();
+      }
+    }
+    return true;
```

Comment:
Note: you should probably just invalidate frames if we have at least one valid (ie: that we return true), if all frame are invalid then we want to do an extra pull.
For example you can look at the current implementation of expand, there is somehow a similar logic (with var `invalid_frames` )